File: src/common/lints/lintSettings.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Lints
open Severity

let ( >>= ) = Base.Result.( >>= )

type 'a t = {
  (* The default value associated with a lint if the lint kind isn't found in the map *)
  default_value: 'a;
  (* Values for lints that have been explicitly set *)
  (* The Loc.t is for settings defined in comments, and is used to find unused lint
   * suppressions. The Loc.t is set to None for settings coming from the flowconfig or --lints.*)
  explicit_values: ('a * Loc.t option) LintMap.t;
}

type warning = int * string

type error = int * string

let default_explicit_values =
  LintMap.singleton Lints.DeprecatedUtility (Severity.Err, None)
  |> LintMap.add Lints.UntypedTypeImport (Severity.Err, None)

let ignored_by_all =
  [
    Lints.DeprecatedUtility;
    Lints.ImplicitInexactObject;
    Lints.AmbiguousObjectType;
    Lints.UninitializedInstanceProperty;
  ]

let config_default kind =
  LintMap.find_opt kind default_explicit_values |> Base.Option.value ~default:(Severity.Off, None)

let of_default default_value =
  let explicit_values = LintMap.of_function ignored_by_all config_default in
  { default_value; explicit_values }

let set_value key value settings =
  let new_map = LintMap.add key value settings.explicit_values in
  { settings with explicit_values = new_map }

let set_all entries settings =
  List.fold_left (fun settings (key, value) -> set_value key value settings) settings entries

let get_default settings = settings.default_value

let get_value lint_kind settings =
  LintMap.find_opt lint_kind settings.explicit_values
  |> Base.Option.value_map ~f:fst ~default:settings.default_value

let get_loc lint_kind settings =
  LintMap.find_opt lint_kind settings.explicit_values |> Base.Option.value_map ~f:snd ~default:None

let is_explicit lint_kind settings = LintMap.mem lint_kind settings.explicit_values

(* Iterate over all lint kinds with an explicit value *)
let iter f settings = LintMap.iter f settings.explicit_values

(* Fold over all lint kinds with an explicit value *)
let fold f settings acc = LintMap.fold f settings.explicit_values acc

(* Map over all lint kinds with an explicit value *)
let map f settings =
  let new_explicit = LintMap.map f settings.explicit_values in
  { settings with explicit_values = new_explicit }

(* SEVERITY-SPECIFIC FUNCTIONS *)

let empty_severities = { default_value = Off; explicit_values = LintMap.empty }

let default_severities = { empty_severities with explicit_values = default_explicit_values }

let is_enabled lint_kind settings =
  match get_value lint_kind settings with
  | Err
  | Warn ->
    true
  | Off -> false

let is_suppressed lint_kind settings = is_enabled lint_kind settings |> not

type parse_result =
  | AllSetting of severity t
  | EntryList of lint_kind list * (severity * Loc.t option)

(* Takes a base LintSettings and a list of labeled lines and returns the corresponding
 * severity LintSettings.t from applying the new lines on top of the base settings if
 * successful. Otherwise, returns an error message along with the label of the
 * line it failed on. *)
let of_lines base_settings =
  let parse_value label value =
    match severity_of_string value with
    | Some severity -> Ok severity
    | None -> Error (label, "Invalid setting encountered. Valid settings are error, warn, and off.")
  in
  let eq_regex = Str.regexp "=" in
  let all_regex = Str.regexp "all" in
  let parse_line (loc, (label, line)) =
    match Str.split_delim eq_regex line with
    | [left; right] ->
      let left = left |> String.trim in
      let right = right |> String.trim in
      parse_value label right >>= fun value ->
      begin
        match left with
        | "all" -> Ok (AllSetting (of_default value))
        | _ ->
          begin
            match kinds_of_string left with
            | Some kinds -> Ok (EntryList (kinds, (value, Some loc)))
            | None -> Error (label, Printf.sprintf "Invalid lint rule \"%s\" encountered." left)
          end
      end
    | _ ->
      Error (label, "Malformed lint rule. Properly formed rules contain a single '=' character.")
  in
  let add_value keys value settings =
    let (new_settings, all_redundant) =
      List.fold_left
        (fun (settings, all_redundant) key ->
          let v = get_value key settings in
          let all_redundant = all_redundant && v = fst value && v <> fst (config_default key) in
          let settings = set_value key value settings in
          (settings, all_redundant))
        (settings, true)
        keys
    in
    if all_redundant then
      Error "Redundant argument. This argument doesn't change any lint settings."
    else
      Ok new_settings
  in
  let rec loop (acc : Severity.severity t) (warnings : warning list) = function
    | [] -> Ok (acc, Base.List.rev warnings)
    | line :: lines ->
      (match parse_line line with
      | Ok (EntryList (keys, value)) ->
        (match add_value keys value acc with
        | Ok acc -> loop acc warnings lines
        | Error msg ->
          let warning = (line |> snd |> fst, msg) in
          loop acc (warning :: warnings) lines)
      | Ok (AllSetting value) ->
        if acc == base_settings then
          loop value warnings lines
        else
          Error
            ( line |> snd |> fst,
              "\"all\" is only allowed as the first setting. Settings are order-sensitive." )
      | Error line_err -> loop acc (line_err :: warnings) lines)
  in
  let loc_of_line line =
    Loc.(
      let start = { line; column = 0 } in
      let _end = { line = line + 1; column = 0 } in
      { source = None; start; _end })
  in
  fun lint_lines ->
    let locate_fun ((label, _) as item) = (loc_of_line label, item) in
    (* Artificially locate the lines to detect unused lines *)
    let located_lines = Base.List.map ~f:locate_fun lint_lines in
    loop base_settings [] located_lines >>= fun (settings, warnings) ->
    let used_locs =
      fold
        (fun _kind (_enabled, loc) acc ->
          Base.Option.value_map loc ~f:(fun loc -> Loc_collections.LocSet.add loc acc) ~default:acc)
        settings
        Loc_collections.LocSet.empty
    in
    let used_locs =
      Base.List.fold
        ~f:(fun acc (line, _warning) -> Loc_collections.LocSet.add (loc_of_line line) acc)
        ~init:used_locs
        warnings
    in
    let first_unused =
      List.fold_left
        (fun acc (art_loc, (label, line)) ->
          match acc with
          | Some _ -> acc
          | None ->
            if
              Loc_collections.LocSet.mem art_loc used_locs
              || Str.string_match all_regex (String.trim line) 0
            then
              None
            else
              Some label)
        None
        located_lines
    in
    let warnings =
      match first_unused with
      | Some label ->
        let warning =
          ( label,
            "Redundant argument. " ^ "The values set by this argument are completely overwritten."
          )
        in
        warning :: warnings
      | None -> warnings
    in
    (* Remove the artificial locations before returning the result *)
    let settings = map (fun (enabled, _loc) -> (enabled, None)) settings in
    Ok (settings, warnings)

let to_string settings =
  let acc = Buffer.create 20 in
  Buffer.add_string acc (Printf.sprintf "all=%s" (settings |> get_default |> string_of_severity));
  iter
    (fun kind (severity, _) ->
      Buffer.add_string
        acc
        (Printf.sprintf ", %s=%s" (string_of_kind kind) (string_of_severity severity)))
    settings;
  Buffer.contents acc

type lint_parse_error =
  | Invalid_setting
  | Malformed_argument
  | Naked_comment
  | Nonexistent_rule
  | Overwritten_argument
  | Redundant_argument

File: rust_port/crates/flow_lint_settings/src/lint_settings.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::collections::HashSet;

use dupe::Dupe;
use flow_parser::loc::Loc;

use crate::lints::DeprecatedTypeKind;
use crate::lints::LintKind;
use crate::severity::Severity;

#[derive(Debug, Clone)]
pub struct LintSettings<T: Copy> {
    // The default value associated with a lint if the lint kind isn't found in the map
    default_value: T,
    // Values for lints that have been explicitly set
    // The Loc is for settings defined in comments, and is used to find unused lint
    // suppressions. The location is set to None for settings coming from the flowconfig or --lints.
    explicit_values: BTreeMap<LintKind, (T, Option<Loc>)>,
}

const IGNORED_BY_ALL: &[LintKind] = &[
    LintKind::ImplicitInexactObject,
    LintKind::AmbiguousObjectType,
    LintKind::UninitializedInstanceProperty,
];

impl<T: Copy> LintSettings<T> {
    fn default_explicit_values() -> BTreeMap<LintKind, (Severity, Option<Loc>)> {
        let mut map = BTreeMap::new();
        map.insert(
            LintKind::DeprecatedType(DeprecatedTypeKind::Bool),
            (Severity::Err, None),
        );
        map.insert(LintKind::InternalType, (Severity::Err, None));
        map.insert(LintKind::LibdefOverride, (Severity::Err, None));
        map.insert(LintKind::NestedComponent, (Severity::Err, None));
        map.insert(LintKind::NestedHook, (Severity::Err, None));
        map.insert(LintKind::UnsafeObjectAssign, (Severity::Err, None));
        map.insert(LintKind::UntypedTypeImport, (Severity::Err, None));
        map
    }

    fn config_default(kind: LintKind) -> (Severity, Option<Loc>) {
        Self::default_explicit_values()
            .get(&kind)
            .cloned()
            .unwrap_or((Severity::Off, None))
    }

    pub fn set_value(&mut self, key: LintKind, value: (T, Option<Loc>)) {
        self.explicit_values.insert(key, value);
    }

    pub fn set_all(&mut self, entries: Vec<(LintKind, (T, Option<Loc>))>) {
        for (key, value) in entries {
            self.set_value(key, value);
        }
    }

    pub fn get_default(&self) -> &T {
        &self.default_value
    }

    pub fn get_value(&self, lint_kind: LintKind) -> &T {
        self.explicit_values
            .get(&lint_kind)
            .map(|(v, _)| v)
            .unwrap_or(&self.default_value)
    }

    pub fn get_loc(&self, lint_kind: LintKind) -> Option<Loc> {
        self.explicit_values
            .get(&lint_kind)
            .and_then(|(_, loc)| loc.clone())
    }

    pub fn is_explicit(&self, lint_kind: LintKind) -> bool {
        self.explicit_values.contains_key(&lint_kind)
    }

    pub fn iter<F>(&self, mut f: F)
    where
        F: FnMut(&LintKind, &(T, Option<Loc>)),
    {
        self.explicit_values.iter().for_each(|(k, v)| f(k, v));
    }

    pub fn fold<F, Acc>(&self, mut f: F, init: Acc) -> Acc
    where
        F: FnMut(LintKind, &(T, Option<Loc>), Acc) -> Acc,
    {
        self.explicit_values
            .iter()
            .fold(init, |acc, (&kind, value)| f(kind, value, acc))
    }

    pub fn replace_with_mapped<F>(&mut self, f: F)
    where
        F: Fn(&(T, Option<Loc>)) -> (T, Option<Loc>),
    {
        self.explicit_values = self
            .explicit_values
            .iter()
            .map(|(&k, v)| (k, f(v)))
            .collect();
    }

    // SEVERITY-SPECIFIC FUNCTIONS

    pub fn empty_severities() -> LintSettings<Severity> {
        LintSettings {
            default_value: Severity::Off,
            explicit_values: BTreeMap::new(),
        }
    }

    pub fn default_severities() -> LintSettings<Severity> {
        LintSettings {
            default_value: Severity::Off,
            explicit_values: Self::default_explicit_values(),
        }
    }
}

enum ParseResult {
    AllSetting(LintSettings<Severity>),
    EntryList(Vec<LintKind>, (Severity, Option<Loc>)),
}

impl Default for LintSettings<Severity> {
    fn default() -> Self {
        Self::empty_severities()
    }
}

impl LintSettings<Severity> {
    pub fn of_default(default_value: Severity) -> LintSettings<Severity> {
        let mut explicit_values = BTreeMap::new();
        for kind in IGNORED_BY_ALL {
            explicit_values.insert(*kind, (Self::config_default(*kind).0, None));
        }
        LintSettings {
            default_value,
            explicit_values,
        }
    }

    pub fn is_enabled(&self, lint_kind: LintKind) -> bool {
        match self.get_value(lint_kind) {
            Severity::Err | Severity::Warn => true,
            Severity::Off => false,
        }
    }

    /// Takes a base LintSettings and a list of labeled lines and returns the corresponding
    /// severity LintSettings from applying the new lines on top of the base settings if
    /// successful. Otherwise, returns an error message along with the label of the
    /// line it failed on.
    pub fn of_lines(
        base_settings: LintSettings<Severity>,
        lint_lines: Vec<(u32, String)>,
    ) -> Result<(LintSettings<Severity>, Vec<Warning>), Error> {
        fn parse_value(line: u32, value: &str) -> Result<Severity, Error> {
            match Severity::severity_of_str(value) {
                Some(severity) => Ok(severity),
                None => Err(Error {
                    line,
                    message:
                        "Invalid setting encountered. Valid settings are error, warn, and off."
                            .to_string(),
                }),
            }
        }

        fn parse_line(loc: &Loc, lint_number: u32, line: &str) -> Result<ParseResult, Error> {
            let parts: Vec<&str> = line.splitn(2, '=').collect();
            if parts.len() != 2 {
                return Err(Error {
                    line: lint_number,
                    message:
                        "Malformed lint rule. Properly formed rules contain a single '=' character."
                            .to_string(),
                });
            }

            let left = parts[0].trim();
            let right = parts[1].trim();

            let value = parse_value(lint_number, right)?;

            if left == "all" {
                Ok(ParseResult::AllSetting(LintSettings::of_default(value)))
            } else {
                match LintKind::parse_from_str(left) {
                    Some(kinds) => Ok(ParseResult::EntryList(kinds, (value, Some(loc.clone())))),
                    None => Err(Error {
                        line: lint_number,
                        message: format!("Invalid lint rule \"{}\" encountered.", left),
                    }),
                }
            }
        }

        fn add_value(
            keys: Vec<LintKind>,
            value: (Severity, Option<Loc>),
            settings: &mut LintSettings<Severity>,
        ) -> Result<(), String> {
            let all_redundant = keys.iter().copied().all(|key| {
                let v = settings.get_value(key);
                v == &value.0 && v != &LintSettings::<Severity>::config_default(key).0
            });

            if all_redundant {
                Err(
                    "Redundant argument. This argument doesn't change any lint settings."
                        .to_string(),
                )
            } else {
                for key in keys {
                    settings.set_value(key, value.dupe());
                }
                Ok(())
            }
        }

        fn loc_of_line(line: u32) -> Loc {
            Loc::mk(None, line as i32, 0, line as i32 + 1, 0)
        }

        // Artificially locate the lines to detect unused lines
        let located_lines: Vec<(Loc, (u32, String))> = lint_lines
            .into_iter()
            .map(|(line_number, line)| (loc_of_line(line_number), (line_number, line)))
            .collect();

        let mut settings = base_settings;
        let mut warnings = Vec::new();
        let mut changed = false;
        for (loc, (lint_number, line)) in &located_lines {
            match parse_line(loc, *lint_number, line) {
                Ok(ParseResult::EntryList(keys, value)) => {
                    match add_value(keys, value, &mut settings) {
                        Ok(()) => {}
                        Err(msg) => warnings.push(Warning {
                            line: *lint_number,
                            message: msg,
                        }),
                    }
                    changed = true;
                }
                Ok(ParseResult::AllSetting(value)) => {
                    if changed {
                        return Err(Error {
                            line: *lint_number,
                            message: "\"all\" is only allowed as the first setting. Settings are order-sensitive."
                                .to_owned(),
                        });
                    }
                    settings = value;
                    changed = true;
                }
                Err(Error {
                    line: l,
                    message: m,
                }) => warnings.push(Warning {
                    line: l,
                    message: m,
                }),
            }
        }

        let mut used_locs = HashSet::new();
        settings.iter(|_kind, (_enabled, loc)| {
            if let Some(loc) = loc {
                used_locs.insert(loc.dupe());
            }
        });
        for Warning {
            line,
            message: _warning,
        } in &warnings
        {
            used_locs.insert(loc_of_line(*line));
        }
        let mut first_unused: Option<u32> = None;
        for (art_loc, (label, line)) in &located_lines {
            if first_unused.is_some() {
                break;
            }
            let trimmed = line.trim();
            if !used_locs.contains(art_loc) && !trimmed.starts_with("all") {
                first_unused = Some(*label);
            }
        }
        if let Some(label) = first_unused {
            let warning = Warning {
                line: label,
                message: "Redundant argument. The values set by this argument are completely overwritten."
                    .to_owned(),
            };
            warnings.insert(0, warning);
        }
        settings.replace_with_mapped(|(enabled, _loc)| (*enabled, None));

        Ok((settings, warnings))
    }

    pub fn to_debug_string(&self) -> String {
        let mut s = format!("all={}", self.get_default().as_str());
        self.iter(|kind, (severity, _)| {
            s.push_str(&format!(", {}={}", kind.as_str(), severity.as_str()));
        });
        s
    }
}

pub struct Warning {
    pub line: u32,
    pub message: String,
}

pub struct Error {
    pub line: u32,
    pub message: String,
}

#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    Hash,
    PartialOrd,
    Ord,
    serde::Serialize,
    serde::Deserialize
)]
pub enum LintParseError {
    InvalidSetting,
    MalformedArgument,
    NakedComment,
    NonexistentRule,
    OverwrittenArgument,
    RedundantArgument,
}
